content-keys-test-vectors.md

```diff
@@ -0,0 +1,117 @@
+# Content Keys Test Vectors
+
+This document contains test vectors for content keys to help implementations conform to the specification. 
+
+## Content Keys Encodings
+
+These test vectors verify the proper SSZ encoding and decoding of content keys
+as specified in the history and state network specs.
+
+### AccountTrieNode
+
+#### Input Parameters
+```
+path = [1,2,0,1]
+node_hash = 0xb8be7903aee73b8f6a59cd44a1f52c62148e1f376c0dfa1f5f773a98666efc2b
+state_root = 0xd1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
+```
+
+#### Expected Output
+```
+content_key = 0x0044000000b8be7903aee73b8f6a59cd44a1f52c62148e1f376c0dfa1f5f773a98666efc2bd1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d01020001
+```
+
+### ContractStorageTrieNode
+
+#### Input Parameters
+```
+address = 0x829bd824b016326a401d083b33d092293333a830
+path = [1, 0, 15, 14, 12, 0]
+node_hash = 0x3e190b68719aecbcb28ed2271014dd25f2aa633184988eb414189ce0899cade5
+state_root = 0xd1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
+```
+
+#### Expected Output
+```
+content_key = 0x01829bd824b016326a401d083b33d092293333a830580000003e190b68719aecbcb28ed2271014dd25f2aa633184988eb414189ce0899cade5d1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d01000f0e0c00
+```
+
+### AccountTrieProof
+
+#### Input Parameters
+```
+address = 0x829bd824b016326a401d083b33d092293333a830
+state_root = 0xd1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d 
+```
+
+#### Expected Output
+```
+content_key = 0x02829bd824b016326a401d083b33d092293333a830d1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
+```
+
+### ContractStorageTrieProof
+
+#### Input Parameters
+```
+address = 0x829bd824b016326a401d083b33d092293333a830
+slot = U256(239304)
+state_root = 0xd1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
+```
+
+#### Expected Output
+```
+content_key = 0x03829bd824b016326a401d083b33d092293333a830c8a6030000000000000000000000000000000000000000000000000000000000d1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
+```
+
+### ContractBytecode
+
+#### Input Parameters
+```
+address = 0x829bd824b016326a401d083b33d092293333a830
+code_hash = 0xd1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
+```
+
+#### Expected Output
+```
+content_key = 0x04829bd824b016326a401d083b33d092293333a830d1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
+```
+
+### HeaderKey
+
+#### Input Parameters
+```
+chain_id = 15
+block_hash = 0xd1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
+```
+
+#### Expected Output
+```
+content_key = 0x010f00d1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
+```
```

Comment:
```suggestion
```content_id = 0xf883705bc68dcd46f9064e8c18dc1c1ae90ba88149b1177bdf1986272779675b
```
```

Reply:
This is a minor organizational issue, but if we put the `content_id` results right underneath the key results, this makes it more concise since we don't have to respecify the inputs and the reader can just glance up to see the inputs. It it not as clean however imo as having a completely separate section for "key to id test vectors" vs "encoding/decoding vectors"

Reply:
That's fine.  I don't have a strong opinion either way.  I'll update the content IDs to match now that we've settled on redoing the content-Key as a SSZ union since that impacts the input to the hash and then you can structure how you want.  My tests have them all lumped together as that made the most sense to me.

Reply:
Regarding the `content-id`, I actually got a different one, assuming the hex string you gave above was a result of taking the resulting u256 value, converting it to a byte array, and then hex encoding it. I think it makes a difference whether it is hash(key) vs hash(serialized_key) since the serialized one has offset bytes in it, changing the input. Either way, I got a result of:

hash(raw_key): 29429819488426549164963845897024823308610593491897334245125007239038426786481
hash(serialized_key): 37312724842171933545395799780116690342117774772998923407358080374963771394635

which did not match the U256 you gave:
15025167517633317571792618561170587584740338038067807801482118109695980329625

There is a possible chance I am using a different keccak function as well.

Reply:
Javascript doesn't support u256 natively so I'm doing it differently though I'm not sure if that has any impact.  Theoretically converting between bytes, hex strings, and u256 shouldn't matter.  I'm taking the raw bytes array from the serialized key and running that directly through the hash function as here - [SHA256](https://github.com/ethereumjs/ultralight/blob/2eb944d53eafafa45d610d346305b5a100e2bbef/packages/portalnetwork/test/historySubnetwork/types.spec.ts#L11) since our hash function takes a bytes array as input and then converting that to a hex string.  I'll try the SHA256 method from one of the other JS web libraries and see if produces a different result (though it shouldn't if they're implemented the same way).

Reply:
I think maybe making the spec a little clearer might help. I looked back at my code and see that there are a few ways that "hash of key" might be interpreted. It seems that `H(key)` vs `H(serialized_key)` gives the same output, but when we say hash of serialized key we might want to specify `H(hexEncode(serialized_key))` because what I had interpreted as hashing the key originally was the raw concatenation of the fields of the key, ie, just concatenating the raw bytes and using that as input w/o any hex encoding.

Having made the key hex encoded, I had the same input string as you had 0x000f00... but still got a different output. I have been using [keccak](https://docs.rs/sha3/0.10.0/sha3/) hash function, from the sha3 family, not sha-256 from the sha2 family, which I believe is what you are using. Maybe try https://www.npmjs.com/package/sha3?

Should we change the spec to use keccak256?

Reply:
The history network spec says specifically to use SHA-256 to derive the `Content-ID` - https://github.com/ethereum/portal-network-specs/blob/master/history-network.md#content-id

I don't have an opinion about sha256 vs keccak but it would seem that at least an initial issue your code isn't following the current spec.

Reply:
But, I'm happy to switch if it makes sense to do so.